### PR TITLE
Add TFPark Estimator API

### PR DESCRIPTION
**Pull request by yangw1234**
- [x] Train
- [x] evaluate
- [x] predict
- [x] example
- [x] unit tests

### CODE REVIEW

File: pyzoo/zoo/tfpark/estimator.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+    return tf.estimator.EstimatorSpec(mode,
+                                      spec.predictions,
+                                      spec.loss,
+                                      train_op,
+                                      spec.eval_metric_ops,
+                                      spec.export_outputs)
+
+
+class EstimatorSpec:
```
Comment (jason-dai) [inline pyzoo/zoo/tfpark/estimator.py]: the name can potentially conflict with `tf.estimator.EstimatorSpec`?

File: pyzoo/zoo/tfpark/estimator.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+    def __init__(self, mode, predictions=None, loss=None, export_outputs=None):
+        self.mode = mode
+        self.predictions = predictions
+        self.loss = loss
+        self.export_outputs = export_outputs
+
+
+class Estimator(object):
```
Comment (jason-dai) [inline pyzoo/zoo/tfpark/estimator.py]: the name can potentially conflict with `tf.estimator.Estimator`?

File: pyzoo/zoo/tfpark/estimator.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+    def __init__(self, model_fn, optimizer, model_dir=None, config=None,
```
Comment (jason-dai) [inline pyzoo/zoo/tfpark/estimator.py]: `optimizer` should be optional?

File: pyzoo/zoo/tfpark/estimator.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            kwargs['mode'] = mode
+        if 'params' in model_fn_args:
+            kwargs['params'] = self.params
+        if 'config' in model_fn_args:
+            kwargs['config'] = config
+
+        model_fn_results = self._model_fn(features=features, **kwargs)
+
+        if not isinstance(model_fn_results, EstimatorSpec):
+            raise ValueError('model_fn should return an EstimatorSpec.')
```
Comment (jason-dai) [inline pyzoo/zoo/tfpark/estimator.py]: this can be confusing if `tf.estimator.EstimatorSpec` is returned

File: pyzoo/zoo/tfpark/estimator.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+        return model_fn_results
+
+    def train(self, input_fn, hooks=None, steps=None, max_steps=None, saving_listeners=None):
```
Comment (jason-dai) [inline pyzoo/zoo/tfpark/estimator.py]: Do we want to support `hooks` and `saving_listeners`? They are not supported in `TFOptimizer`; and also we do not use `max_steps` for `TFOptimizer` below.

File: pyzoo/zoo/tfpark/estimator.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+        with tf.Graph().as_default() as g:
+            global_step_tensor = self.estimator._create_and_assert_global_step(g)
+            add_step_input = tf.placeholder(dtype=tf.int64, shape=())
+            assign_step = tf.assign_add(global_step_tensor, add_step_input)
```
Comment (jason-dai) [inline pyzoo/zoo/tfpark/estimator.py]: `global_step_tensor` is not used for `TFOptimizer` below?

**Comment by yangw1234** [comment]
http://172.168.2.101:8080/view/ZOO-PR/job/ZOO-PR-Validation/ jenkins passed except one failure based by "no space left on device"

**Comment by yangw1234** [comment]
merge it for now

File: pyzoo/zoo/tfpark/estimator.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+            raise ValueError(
+                'model_fn does not take labels, but input_fn returns labels.')
+    if 'mode' in model_fn_args:
+        kwargs['mode'] = mode
+    if 'params' in model_fn_args:
+        kwargs['params'] = params
+    if 'config' in model_fn_args:
+        kwargs['config'] = config
+
+    spec = model_fn(features=features, **kwargs)
```
Comment (jason-dai) [inline pyzoo/zoo/tfpark/estimator.py]: spec = self._call_model_fn(...)?

File: pyzoo/zoo/tfpark/estimator.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+        grads_and_vars = optimizer.compute_gradients(spec.loss)
+
+        vars_with_grad = [v for g, v in grads_and_vars if g is not None]
+        if not vars_with_grad:
+            raise ValueError(
+                "No gradients provided for any variable, check your graph for ops"
+                " that do not support gradients, between variables %s and loss %s." %
+                ([str(v) for _, v in grads_and_vars], spec.loss))
+
+        train_op = spec.optimizer.apply_gradients(grads_and_vars,
```
Comment (jason-dai) [inline pyzoo/zoo/tfpark/estimator.py]: train_op = optimizer.minimize(loss, global_step=tf.train.get_global_step())?

Comment (yangw1234) [reply pyzoo/zoo/tfpark/estimator.py]: I'll fix that in the next pr.

File: pyzoo/zoo/tfpark/estimator.py (hunk 7 above)
Comment (yangw1234) [reply pyzoo/zoo/tfpark/estimator.py]: I'll fix that in the next pr.

File: pyzoo/zoo/tfpark/estimator.py (hunk 8 above)
Comment (yangw1234) [reply pyzoo/zoo/tfpark/estimator.py]: fixed in #1256
